datamodel: create base structure for datamodel

ppanero opened the pull request:
Create a base structure for the DataCite based datamodel. This aligns with the `records` datamodel inside `cookiecutter-invenio-instance` (previously self-contained in `cookiecutter-invenio-datamodel`).

It contains:
- JSON Schemas
- Elasticsearch mappings
- Loaders
- Serializers
- Deserializers

*NOTE*: There's still a lot of discussion to be had on what should be internal and what metadata, etc. This just aims to have a clear code structure of the datamodel and a starting point to add more fields.

Closes https://github.com/inveniosoftware/invenio-rdm-records/issues/6

ppanero commented:
**NOTE***: I think both commits can be squashed in one

fenekku commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
+        },
+        "publication_date": {
+          "type": "date",
+          "format": "date"
```
normal: Why `date` here and `date-time` in the jsonschema above? (`date` doesn't include time)

ppanero replied:
I'd swear I had answer this, somehow it didn't get through. Sorry.

The [Elasticsearch `date` type](https://www.elastic.co/guide/en/elasticsearch/reference/current/date.html) includes the time. There is actually no `date-time` in ES (AFAIK). But JSON Schema requires `date-time`. That's why the mapping has a different value than the jsonschema.

fenekku commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
+              "type": "object",
+              "properties": {
+                "source": {
+                  "type": "text"
```
`keyword` might be nice here for aggregation purposes and since source will be from a limited set.

ppanero replied:
:+1:

fenekku commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
+              "type": "keyword"
+            },
+            "email": {
+              "type": "text"
```
`keyword` too I think.

fenekku commented on invenio_rdm_records/marshmallow/json.py:
```diff
@@ -0,0 +1,71 @@
+    """Access schema."""
+
+    metadata_restricted = fields.Bool(required=True)
+    files_restricted = fields.Bool(required=True)
```
Made an issue to cover implementation / discussion on this: #20 . Maybe we can put a `# TODO:` on those metadata fields we haven't gone through yet.

ppanero replied:
:+1:

fenekku commented on invenio_rdm_records/marshmallow/json.py:
```diff
@@ -0,0 +1,71 @@
+    owners = fields.List(fields.Integer(),
+                         validate=validate.Length(min=1),
+                         required=True)
+    keywords = fields.List(SanitizedUnicode(), many=True)
```
As above, would skip for now

fenekku replied:
fine with keeping until revisit.

fenekku commented on invenio_rdm_records/marshmallow/json.py:
```diff
@@ -0,0 +1,71 @@
+    revision = fields.Integer(dump_only=True)
+    updated = fields.Str(dump_only=True)
+    links = fields.Dict(dump_only=True)
+    id = PersistentIdentifier(attribute='pid.pid_value')
```
comment: We will have to watch out for this one. I remember it acting tricky with deposits and edits flows.

fenekku commented on tests/conftest.py:
```diff
@@ -33,13 +26,6 @@ def celery_config():
-    return factory
+def create_app():
+    """Create app fixture."""
+    return create_api
```
normal: This is the same `create_app` as `tests/api/conftest.py`. Should we have the following:

`tests/conftest.py`: `create_app` creates an api and ui application
`tests/api/conftest.py`: `create_app` creates an api only application
`tests/ui/conftest.py`: `create_app` creates ui only application

ppanero replied:
:+1: Changed the ones. Missing UI for when we work on UI.

fenekku commented on tests/test_examples_app.py:
```diff
@@ -49,10 +49,3 @@ def example_app():
-    """Test example app."""
-    cmd = 'curl http://0.0.0.0:5000/'
-    output = subprocess.check_output(cmd, shell=True)
-    assert b'Welcome to Invenio-RDM-Records' in output
```
:fire: yes

ppanero commented on invenio_rdm_records/config.py:
```diff
@@ -56,6 +64,50 @@
+    records=dict(
+        aggs=dict(
+            type=dict(terms=dict(field='type')),
+            keywords=dict(terms=dict(field='keywords'))
```
Just the key in the datamodel. https://invenio-records-rest.readthedocs.io/en/latest/configuration.html#invenio_records_rest.config.RECORDS_REST_FACETS

`type` seems to be an inherited mistake from cookiecutter.

fenekku replied:
Oups! Haha, those "self:" comments were not meant to be submitted. I use them to keep track of my own questions (that one was resolved later in the code indeed).

ppanero commented on invenio_rdm_records/config.py:
```diff
@@ -56,6 +64,50 @@
+        ),
+    )
+)
+"""Setup sorting options."""
```
Maybe we can open a ticket? it worked for me in another Invenio app I developed (UI with React-SearchKit though)

fenekku replied:
Exactly, I wasn't using React-SearchKit. If the above works in React-SearchKit and we will be using it, then ignore the comment and this is all good.

ppanero replied:
I think is the idea, but cant be 100% sure until January when UI work starts.

ppanero commented on invenio_rdm_records/ext.py:
```diff
@@ -31,12 +31,20 @@ def init_app(self, app):
+                          'RECORDS_REST_DEFAULT_SORT']:
+                    if k == n and with_endpoints:
+                        app.config.setdefault(n, {})
+                        app.config[n].update(getattr(config, k))
```
:heart_eyes: Nice improvement!

ppanero commented on invenio_rdm_records/jsonschemas/records/record-v1.0.0.json:
```diff
@@ -53,6 +53,76 @@
+        "type": "string"
+      },
+      "type": "array"
+    },
```
i agree, just that the cookiecutter has them. So I added them to be consistent. I would vote to leave it, we will address it anyway sooner or later.

fenekku replied:
Ok, let's leave `keywords` in and address it when we tackle #3.

ppanero commented on invenio_rdm_records/jsonschemas/records/record-v1.0.0.json:
```diff
@@ -53,6 +53,76 @@
+        ]
+
+      }
+    },
```
For now, it is accepted and we change after discussion?

fenekku replied:
Absolutely! Since we have an issue, this is section is good to go and discussions will be had on the RFC / issue.

ppanero commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
         },
         "$schema": {
-          "type": "keyword"
+          "type": "text",
```
I changed it to keyword. However since it has `index: false` it will be treated equally --> Not be treated/indexed :joy:

fenekku replied:
yeah I guess so :laughing:

ppanero commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
+        },
+        "suggest_title": {
+          "type": "completion"
         },
```
Sort: shall we add it to a ticket along with the other sorting issue above?

Suggest_title: Its used for autocompletion https://www.elastic.co/guide/en/elasticsearch/reference/current/search-suggesters.html#completion-suggester (I prefer n-grams since they provide better results. But this are much more efficient)

fenekku replied:
For "title" sorting, we might as well just add it now and save us the overhead of a ticket.

For `suggest_title`, I was actually wondering about the use-case. Where/when are we auto-completing titles?

ppanero replied:
For sorting. It works for me with React Search kit. AFAIK, we are intending to use it, but we wont be sure 100% until January when we start with UI.

Suggestion, I fear I dont see it in Zenodo for titles. I see it for licenses. In other instances (e.g. CDS videos its used already, see screenshot). I think is a nice improvement, shall we leave it?
![image](https://user-images.githubusercontent.com/6756943/67189778-21b8c000-f3ef-11e9-9c61-f712f8e082bc.png)

ppanero commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
+            "email": {
+              "type": "text"
+            },
+            "name": {
```
Is the issue created for this? Or shall I?

fenekku replied:
This discussion would be in the "Contributors" issue, we are good here, no issue to be created.

ppanero commented:
@fenekku addressed all your comments :) Thanks a lot!

Just reminds the `keyword`. I would create an issue and leave as it is for now, to be consistent with cookiecutter (which was the point of the issue). And have a further discussion on the keywords/subjetc(s)/etc. and change as a result in the future. What do you think?

fenekku commented on invenio_rdm_records/mappings/v6/records/record-v1.0.0.json:
```diff
@@ -22,20 +22,60 @@
         },
+        "keywords": {
+          "type": "keyword"
+        },
```
fine with including since we will revisit.

fenekku commented on invenio_rdm_records/templates/records/record.html:
```diff
@@ -0,0 +1,47 @@
+  
+    Invenio-RDM-Records is free software; you can redistribute it and/or modify it
+    under the terms of the MIT License; see LICENSE file for more details.
+-->
```
minor: this one is a Jinja2 template so we were good :D 

(change back if you want to send less bytes over the wire)
